File: 15-3sum/3sum.scala
object Solution {
    def threeSum(nums: Array[Int]): List[List[Int]] = {
        val result = scala.collection.mutable.ListBuffer[List[Int]]()
        val arr = nums.sorted

        for (i <- 0 until arr.length - 2) {
            // Skip duplicates
            if (i > 0 && arr(i) == arr(i - 1)) {
                // continue
            } else {
                var left = i + 1
                var right = arr.length - 1

                while (left < right) {
                    val sum = arr(i) + arr(left) + arr(right)

                    if (sum == 0) {
                        result += List(arr(i), arr(left), arr(right))

                        // Skip duplicates for left
                        while (left < right && arr(left) == arr(left + 1)) left += 1
                        // Skip duplicates for right
                        while (left < right && arr(right) == arr(right - 1)) right -= 1

                        left += 1
                        right -= 1
                    }
                    else if (sum < 0) {
                        left += 1
                    }
                    else {
                        right -= 1
                    }
                }
            }
        }

        result.toList
    }
}

File: 15-3sum/3sum.c
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */

#include <stdlib.h>

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int** threeSum(int* nums, int numsSize, int* returnSize, int** returnColumnSizes) {
    *returnSize = 0;
    if (numsSize < 3) return NULL;

    qsort(nums, numsSize, sizeof(int), compare);

    int capacity = 1000;  // initial capacity
    int** result = (int**)malloc(sizeof(int*) * capacity);
    *returnColumnSizes = (int*)malloc(sizeof(int) * capacity);

    for (int i = 0; i < numsSize - 2; i++) {
        // skip duplicate i
        if (i > 0 && nums[i] == nums[i - 1]) continue;

        int left = i + 1;
        int right = numsSize - 1;

        while (left < right) {
            int sum = nums[i] + nums[left] + nums[right];

            if (sum == 0) {
                // resize if needed
                if (*returnSize >= capacity) {
                    capacity *= 2;
                    result = (int**)realloc(result, sizeof(int*) * capacity);
                    *returnColumnSizes = (int*)realloc(*returnColumnSizes, sizeof(int) * capacity);
                }

                int* triplet = (int*)malloc(sizeof(int) * 3);
                triplet[0] = nums[i];
                triplet[1] = nums[left];
                triplet[2] = nums[right];

                result[*returnSize] = triplet;
                (*returnColumnSizes)[*returnSize] = 3;
                (*returnSize)++;

                // skip duplicates (left)
                while (left < right && nums[left] == nums[left + 1]) left++;
                // skip duplicates (right)
                while (left < right && nums[right] == nums[right - 1]) right--;

                left++;
                right--;
            }
            else if (sum < 0) {
                left++;
            }
            else {
                right--;
            }
        }
    }

    return result;
}